plugins/plugin-testing-java/plugin-testing-base/src/main/java/org/eclipse/che/plugin/java/testing/AbstractJavaTestRunner.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2012-2017 Red Hat, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.plugin.java.testing;
+
+import static java.util.Collections.emptyList;
+import static org.eclipse.che.api.languageserver.service.LanguageServiceUtils.prefixURI;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+
+import java.io.IOException;
+import java.net.Socket;
+import java.util.List;
+import java.util.Random;
+import java.util.stream.Collectors;
+import org.eclipse.che.api.testing.server.framework.TestRunner;
+import org.eclipse.che.api.testing.shared.TestDetectionContext;
+import org.eclipse.che.api.testing.shared.TestExecutionContext;
+import org.eclipse.che.api.testing.shared.TestPosition;
+import org.eclipse.che.plugin.java.languageserver.JavaLanguageServerExtensionManager;
+import org.eclipse.che.plugin.java.languageserver.dto.DtoServerImpls.TestPositionDto;
+
+/**
+ * Abstract java test runner. Can recognize test methods, find java project and compilation unit by
+ * path.
+ */
+public abstract class AbstractJavaTestRunner implements TestRunner {
+  private int debugPort = -1;
+  private JavaLanguageServerExtensionManager extensionManager;
+
+  public AbstractJavaTestRunner(JavaLanguageServerExtensionManager extensionManager) {
+    this.extensionManager = extensionManager;
+  }
+
+  @Override
+  public List<TestPosition> detectTests(TestDetectionContext context) {
+    List<TestPositionDto> testPositionDtos =
+        extensionManager.detectTest(
+            prefixURI(context.getFilePath()), getTestAnnotation(), context.getOffset());
+
+    return testPositionDtos
+        .stream()
+        .map(
+            it ->
+                newDto(TestPosition.class)
+                    .withFrameworkName(getName())
+                    .withTestName(it.getTestName())
+                    .withTestNameLength(it.getTestNameLength())
+                    .withTestBodyLength(it.getTestBodyLength())
+                    .withTestNameStartOffset(it.getTestNameStartOffset()))
+        .collect(Collectors.toList());
+  }
+
+  /** Returns import of test annotation */
+  protected abstract String getTestAnnotation();
```

Comment:
Consider making this a constructor parameter. Favor composition over inheritance.